ideas/apq.h

```diff
@@ -0,0 +1,285 @@
+*                                  connection or recycling operation for pool.
+*/
+using pg_connect_handle = std::shared_ptr<PGconn>; //, std::function<void(PGconn*)>>;
+
+
+/**
+* We can try to use standard error_code, but we will loose the error message
+* from database. But if we will use the custom error_code we are forced to
+* implement our own yield_context and async_result infrastructure. This perspective
+* is not attractive. We can provide some kind of yield_context support, but with
+* message loss in case of error_code bound to yield (i.e. yield[ec])
+*/
+using error_code = boost::system::error_code; // std:: / custom ?
+
+/**
+* Connection - represents binding of hative handle an io_service (io_context) object
+* Should we allow to copy connection itself, or the object must be move-only?
+*/
+struct connection {
+	io_service& ios;
+	pg_connect_handle h;
+
+	// Connection can be ConnectionProvider, so it provides itself
+	template <typename Handler>
+	friend void get_connection_impl(connection c, Handler&& h) const {
+		asio_handler_invoke([&]{ h(error_code{}, c);}, std::addressof(h));
+	}
+};
+
+/**
+* ConnectionProvider - concept provides applicable get_connection(Handler) function.
+* Handler - concept of a callback with signature (error_code, connection). Connection
+* is described below
+*/
+
+namespace detail {
+
+template <typename ConnectionProvider, typename Handler>
+void get_connection_impl(ConnectionProvider&& p, Handler&& h) {
+	p(std::forward<Handler>(h));
+}
+
+} // detail
+
+/**
+* Function to get connection from provider. This customization point allows
+* us to work with different kinds of connection providing. E.g. single
+* connection, get connection from pool, lazy connection, retriable connection
+* and so on.
+*/
+template <typename ConnectionProvider, typename CompletionToken>
+auto get_connection(ConnectionProvider&& p, CompletionToken&& token) {
+	using detail::get_connection_impl;
+	detail::async_result_init<CompletionToken, void (error_code, connection)> init(std::move(token));
+	get_connection_impl(std::forward<ConnectionProvider>(p), init.handler);
+	return init.result.get();
+}
+
+
+struct dsn; // or just an alias to std::string?
+
+
+class connection_pool {
+public:
+	template <typename Handler>
+	void get_connection(io_service& ios, Handler&& h);
+};
+
+// Returns ConnectionProvider for specified dsn
+auto make_connection_provider(io_service&, dsn); 
+
+// Return ConnectionProvider for specified connection pool
+auto make_connection_provider(io_service& ios, connection_pool& pool) {
+	return [](auto&& handler) {
+		pool.get_connection(ios, std::move(handler));
+	}
+}
+
+enum class parameters_transfer_mode {binary, text};
+
+template <typename T>
+using cursor = std::unique_ptr<void(T&, void (error_code, size_type))>;
+
+class transaction_connection;
+
+/**
+* Represents raw data row in case of request.
+*/
+class data_row;
```

Comment:
Why not just ```row```?

Reply:
I can't remember the actual reasons why not - let it be just ```row``` :)